### ISSUE TITLE
Feature: Status enum on PrEntity

### ISSUE THREAD

**Issue by Layla-P** [Layla-P/HacktoberfestProject#14]
** Description: **

PRs marked "invalid" won't count so we will need to be able to parse through all PRs as a batch job and mark their status depending on the status of the PR.

** Steps: **

1. Create a PrStatusEnum class
2. Statuses such as
```
Default,
Pending,
Accepted,
Invalid
```
3. Update the insert method to set this as pending on record creation.

**Comment by CrypticEngima** [comment]
@Layla-P I am currently struggling to lockdown in the Github API how I can aquire this information to implament this functionality locally.

**Comment by rickvdbosch** [comment]
The GitHub webhooks are also [available for Pull Requests](https://docs.github.com/en/free-pro-team@latest/developers/webhooks-and-events/webhook-events-and-payloads#pull_request). I can imagine we cannot create webhooks in the *original* repository since that's not owned by the GitHub user registering with the HacktoberfestProject. 

But if it also triggers for PRs created *out of the forked repo*, we could use those. And then it's very simple to create an Azure Function that updates the PR in our Table Storage. Might be worth having a look at.

**Comment by Layla-P** [comment]
When we make a call to the PR API it returns an object including the status.
https://docs.github.com/en/free-pro-team@latest/rest/reference/pulls#get-a-pull-request

I think it is the `state` - as long as that isn't marked as `invalid` it should contribute to their logged PRs I believe.

So when a new PR is logged, we set the status to pending as we won't have checked the state of the PR yet.
Then at the end of Hacktoberfest, I see us doing a webjob to parse the data and update the status depending on the state of the PR.

Does that make sense? Happy to hop on a call :-)

**Comment by CrypticEngima** [comment]
Just had a look at the state enum of a returned Pr and this is all it has 

```
	// Summary:
	//     Items that are open OR closed
	public enum ItemState
	{
		//
		// Summary:
		//     Items that are open
		Open = 0,
		//
		// Summary:
		//     Items that are closed
		Closed = 1
	}
```

**Comment by SiliconOrchid** [comment]
@Layla-P has indicated that records should be processed in a one-time operation at the end of HacktoberFest

At risk of over-complicating ... is there anything to be gained by processing this more frequently?   Perhaps to provide up-to-date feedback [on the status of their PR submissions] for users returning to the site.


If this is something that is desirable;, we need to address the update mechanism.

For example:

* Re-occurring scheduled batch process - e.g. WebJob on a CRONjob timer.
    * A webjob will need to have an additional console project to be created.
* Event-driven from the GitHub end? (speculating that there is a webhook available at GitHub API - I have not checked this even exists).

**Comment by SiliconOrchid** [comment]
We may need to consider having an additional status to indicate "PR out of date range" (to be able to flag PR submissions that do not fall within the competition date range of the month of October.)

This line of thought may also lead us down path of having multiple status enums/codes that   - instead of just simply "accepted", "rejected" or "pending".

**Comment by CrypticEngima** [comment]
@Layla-P  is this a must have for MVP or a nice to have just aware of deadline ?

**Comment by Layla-P** [comment]
@CrypticEngima not essential for MVP we can work on this next week - just may want the status in the entity models that @rickvdbosch is working on

**Comment by SiliconOrchid** [comment]
I have attempted to understand the Enum values for GitHub PR "Status".

Tagging @CrypticEngima as he has been looking at this.

---

Referring to this doc:
[https://developer.github.com/v3/pulls/reviews/#list-reviews-for-a-pull-request](https://developer.github.com/v3/pulls/reviews/#list-reviews-for-a-pull-request)

The JSON response has a key "state" with associated string values

I couldn't see an explicit listing of possible values,  but I've noted the following can be found in the examples of  json response:

"APPROVED"
"PENDING"
"CHANGES_REQUESTED"
"DISMISSED"

Note the above related to v3 API

---

Confusingly, for v4 API, the following resource:
[https://developer.github.com/v4/enum/pullrequeststate/](https://developer.github.com/v4/enum/pullrequeststate/)

... which was linked-to from [https://developer.github.com/v4/object/pullrequest/](https://developer.github.com/v4/object/pullrequest/) ...

... indicates that the key field "State", has values of type "PullRequestState" - but there are just three values:  

"CLOSED" 
"MERGED"
"OPEN"

**Comment by CrypticEngima** [comment]
Just ran a test with a closed pr on my test repo and the state comes back as closed. am just looking at status url info

**Comment by Layla-P** [comment]
Update: https://hacktoberfest.digitalocean.com/hacktoberfest-update

```
PRs count if:
Submitted in a repo with the hacktoberfest topic AND
during the month of October AND (
  The PR is merged OR
  The PR is labelled as hacktoberfest-accepted by a maintainer OR
  The PR has been approved
)

```
According to the new rules on Hacktoberfest for anti-spam we'll need to check for the above points.

**Comment by CrypticEngima** [comment]
@Layla-P  I think using these rules alone will allow us to created the status of the PR' s do you want me to run the check on retrive of a Pr (So it consistlntly upto date) and add a status that can be used to key off of for the front end to display say colors on the PR to show their applicability to Hacktoberfest. I don't think we should save this value as it needs to be acurate to current state of remote repo at time of showing in my opinion.

**Comment by rickvdbosch** [comment]
I think it might be an optimization to *not store* the status as long as a PR does not yet count according to the rules. As soon as a PR counts we *should store* this, so we don't have to ask GitHub for all of the PRs every time.

**Comment by CrypticEngima** [comment]
@rickvdbosch  Good point i will look into this
